ServerCore/Pages/Threads/PuzzleThreads.cshtml.cs

```diff
@@ -149,20 +149,34 @@ public PuzzleThreadsModel(PuzzleServerContext serverContext, UserManager<Identit
                 throw new NotSupportedException($"EventRole [{EventRole}] is not supported in PuzzleThreads");
             }
 
-            // Filter down to only latest messages
-            LatestMessagesFromEachThread = await messages
-                .GroupBy(message => message.ThreadId)
-                .Select(group => group.OrderByDescending(message => message.CreatedDateTimeInUtc).First())
-                .ToListAsync();
+            // Filter down to only latest messages in memory
+            List<Message> retrievedMessages = await messages.ToListAsync();
+            var threadIdToMostRecentMessage = new Dictionary<string, Message>();
+            foreach (Message message in retrievedMessages)
+            {
+                if (!threadIdToMostRecentMessage.ContainsKey(message.ThreadId))
+                {
+                    threadIdToMostRecentMessage.Add(message.ThreadId, message);
+                }
+                else if (message.CreatedDateTimeInUtc > threadIdToMostRecentMessage[message.ThreadId].CreatedDateTimeInUtc)
+                {
+                    threadIdToMostRecentMessage[message.ThreadId] = message;
+                }
+            }
+
+            IEnumerable<Message> finalFilteredMessages = threadIdToMostRecentMessage.Values
+                .AsEnumerable()
+                .OrderByDescending(Message => Message.CreatedDateTimeInUtc);
```

Comment:
If you OrderByDescending in the original query, then the foreach above can be simplified by removing the else clause altogether.

Reply:
The original issue was that we were grouping and sorting, which lead to that massive block.
This was the way of removing the grouping and sorting.